widgets/pkg_all.lua

```diff
@@ -33,16 +33,34 @@ local function worker(format, warg)
     }
 
     -- Check if updates are available
+    local function parse(str, skiprows)
+        local size, lines, first = 0, "", skiprows or 0
+        for line in str:gmatch("[^\r\n]+") do
+            if size >= first then
+                lines = lines .. (size == first and "" or "\n") .. line
+            end
+            size = size + 1
+        end
+        size = math.max(size-first, 0)
+        return {size, lines}
+    end
+    
+    -- Select command
     local _pkg = manager[warg]
-    local f = io.popen(_pkg.cmd)
+    spawn.easy_async(_pkg.cmd, function(stdout) callback(parse(stdout, _pkg.sub)) end)
+end
+-- }}}
 
-    for line in f:lines() do
-        updates = updates + 1
-    end
-    f:close()
+-- {{{ Packages widget type
+local function worker(format, warg)
+    local ret = nil
+    
+    pkg_all.async(warg, function(data) ret = data end)
 
-    return {_pkg.sub and math.max(updates-_pkg.sub, 0) or updates}
+    while ret==nil do end
```

Comment:
This is just to maintain the old behaviour of the `worker` function, but it is never actually called.
I think the best would be to have only the function async for all the widgets (or change the `worker` functions to take a callback as argument). Like that all the widgets are asynchronous. It should not be too complicated to implement: with a function `parse` when io.popen was used, or calling directly `callback(...)` at the end if all the code is synchronous.
The only problem is that the API would change, so all the widgets have to be rewritten...